internal/controllers/topology/cluster/structuredmerge/patchhelper.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package structuredmerge
+
+import (
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+const topologyManager = "topology"
+
+type serverSidePatchHelper struct {
+	client         client.Client
+	modified       *unstructured.Unstructured
+	hasChanges     bool
+	hasSpecChanges bool
+}
+
+// NewServerSidePatchHelper returns a new PatchHelper using server side apply.
+func NewServerSidePatchHelper(scheme *runtime.Scheme, original, modified client.Object, c client.Client, opts ...HelperOption) (PatchHelper, error) {
+	helperOptions := &HelperOptions{}
+	helperOptions = helperOptions.ApplyOptions(opts)
+	helperOptions.allowedPaths = []contract.Path{
+		// apiVersion, kind, name and namespace are required field for a server side apply intent.
+		{"apiVersion"},
+		{"kind"},
+		{"metadata", "name"},
+		{"metadata", "namespace"},
+		// the topology controller controls/has an opinion for labels, annotation and spec only.
+		{"metadata", "labels"},
+		{"metadata", "annotations"},
+		{"spec"},
+	}
+
+	// If required, convert the original and modified objects to unstructured and filter out all the information
+	// not relevant for the topology controller.
+
+	var originalUnstructured *unstructured.Unstructured
+	if original != nil {
+		originalUnstructured = &unstructured.Unstructured{}
+		switch original.(type) {
+		case *unstructured.Unstructured:
+			originalUnstructured = original.DeepCopyObject().(*unstructured.Unstructured)
+		default:
+			if err := scheme.Convert(original, originalUnstructured, nil); err != nil {
```

Comment:
```suggestion
			if err := scheme.Convert(original.DeepCopy(), originalUnstructured, nil); err != nil {
```
We need to use deepcopy, otherwise filterObject changes `original`.

Reply:
@chrischdi can we recheck this?
filterObject acts on the unstructured object, not on the typed one

Reply:
I tested it again: should not be necessary anymore due to the change from the Generic types to the typed ones 👍